libraries/state_history/log.cpp

```diff
@@ -288,52 +314,124 @@ void state_history_log::split_log() {
+      }
+      block_num_type block_num = chain::block_header::num_from_id(id);
+      auto shared_data = std::make_shared<std::vector<char>>(std::move(data));
+      boost::asio::post(work_strand, [this, entry = log_entry_type{id, prev_id, shared_data} ]() {
+         std::unique_lock<std::mutex> lock(this->mx);
+         write_entry(entry, lock);
+      });
+      cached[block_num] = shared_data;
+
+      while (cached.size() > num_buffered_entries && cached.begin()->second.use_count() == 1) {
+         // the log entry has been removed from write_queue, so we can safely erase it. 
+         cached.erase(cached.begin());
+      }
 
-chain::bytes state_history_traces_log::get_log_entry(block_num_type block_num) {
+      fc_ilog(logger,"store_entry name=${name}, block_num=${block_num} cached.size = ${sz}, num_buffered_entries=${num_buffered_entries}, id=${id}",
+            ("name", name)("block_num", block_num)("sz", cached.size())("num_buffered_entries", num_buffered_entries)("id", id));
 
-   auto get_traces_bin = [block_num](auto& ds, uint32_t version, std::size_t size) {
-      auto start_pos = ds.tellp();
+   }
+   else {
+      state_history_log_header header{.magic = ship_magic(ship_current_version), .block_id = id};
+      auto [block_num, start_pos] = write_entry_header(header, prev_id);
       try {
-         if (version == 0) {
-            return state_history::zlib_decompress(ds);
-         }
-         else {
-            std::vector<state_history::transaction_trace> traces;
-            state_history::trace_converter::unpack(ds, traces);
-            return fc::raw::pack(traces);
+         this->write_payload(write_log, data);
+         write_entry_position(header, start_pos, block_num);
+      } catch (...) {
+         write_log.close();
+         boost::filesystem::resize_file(write_log.get_file_path(), start_pos);
+         write_log.open("rb+");
+         throw;
+      }
+   }
+}
+
+void state_history_log::write_entry(const log_entry_type& entry, std::unique_lock<std::mutex>& lock) {
+   state_history_log_header header{.magic = ship_magic(ship_current_version), .block_id = entry.id};
+   auto [block_num, start_pos] = write_entry_header(header, entry.prev_id);
+   try {
+      lock.unlock();
+      this->write_payload(write_log, *entry.data);
+      lock.lock();
+      write_entry_position(header, start_pos, block_num);
+      fc_dlog(logger,"entry block_num=${block_num} id=${id} written", ("block_num", block_num) ("id", entry.id));
+   } catch (...) {
+      write_log.close();
+      boost::filesystem::resize_file(write_log.get_file_path(), start_pos);
+      write_log.open("rb+");
+      throw;
+   }
+}
+
+state_history_traces_log::state_history_traces_log(const state_history_config& config)
+    : state_history_log("trace_history", config) {
+}
+
+std::shared_ptr<std::vector<char>> state_history_traces_log::get_log_entry(block_num_type block_num) {
+
+   auto get_entry_from_disk = [&] {
```

Comment:
This seems rather convoluted. Is there a reason you create this large lambda. Seems like you could just add:

```
   if (thr.joinable()) {
      auto itr = cached.find(block_num);
      if (itr != cached.end()) {
         std::vector<state_history::transaction_trace> traces;
         auto ds = fc::datastream<const char*>(itr->second->data(), itr->second->size());
         state_history::trace_converter::unpack(ds, traces);
         return std::make_shared<std::vector<char>>(fc::raw::pack(traces));
      }
  }
```

As the first lines of `get_log_entry` and not need the large lambda.